### PR TITLE
feat: remove counting withdrawals against user limits

### PR DESCRIPTION
**Pull request by ramenforbreakfast**
This PR focuses on changing the `CollateralDepositRecord` contract to not count withdrawals against user deposit totals.

In addition, without trying to change the test case too much, I made some edits to it since some things we originally did, did not make sense to me (e.g. using an arbitrary `testDepositToWithdrawFrom` when we already had a `TEST_AMOUNT_ONE` and `TEST_AMOUNT_TWO` to use).

### CODE REVIEW

File: apps/smart-contracts/core/test/CollateralDepositRecord.test.ts [hunk 1]
```diff
@@ -119,51 +120,47 @@ describe('=> CollateralDepositRecord', () => {
...
+        userDepositsBeforeSecondWithdrawal
       )
     })
 
-    it('should set the deposit total to zero instead of underflowing if withdrawal amount is greater than the existing total', async () => {
-      expect(await depositRecord.getGlobalNetDepositAmount()).to.be.eq(testDepositToWithdrawFrom)
-      expect(await depositRecord.getUserDepositAmount(user.address)).to.be.eq(
-        testDepositToWithdrawFrom
-      )
+    it('should set global total to zero instead of underflowing if withdrawal amount is greater than existing total', async () => {
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/CollateralDepositRecord.test.ts:154]: nit: start using new naming convention if you're updating any tests
```suggestion
    it('sets global deposits to 0 if withdrawal > global deposits', async () => {
```

File: apps/smart-contracts/core/test/CollateralDepositRecord.test.ts [hunk 2]
```diff
@@ -119,51 +120,47 @@ describe('=> CollateralDepositRecord', () => {
...
+      const globalDepositsBefore = await depositRecord.getGlobalNetDepositAmount()
+      expect(globalDepositsBefore).to.be.gt(0)
+      const userDepositsBefore = await depositRecord.getUserDepositAmount(user.address)
+      expect(userDepositsBefore).to.be.gt(0)
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/CollateralDepositRecord.test.ts:158]: the user deposits part should be a separate `leaves user deposit unchanged if withdrawal > global net deposits` test

File: apps/smart-contracts/core/test/CollateralDepositRecord.test.ts [hunk 3]
```diff
@@ -119,51 +120,47 @@ describe('=> CollateralDepositRecord', () => {
...
-      await depositRecord.connect(user).recordWithdrawal(user.address, TEST_AMOUNT_ONE)
+      await depositRecord.connect(user).recordWithdrawal(user.address, TEST_AMOUNT_TWO)
 
       expect(await depositRecord.getGlobalNetDepositAmount()).to.eq(
-        globalDepositsBeforeSecondWithdrawal.sub(TEST_AMOUNT_ONE)
+        globalDepositsBeforeSecondWithdrawal.sub(TEST_AMOUNT_TWO)
       )
       expect(await depositRecord.getUserDepositAmount(user.address)).to.eq(
-        accountDepositsBeforeSecondWithdrawal.sub(TEST_AMOUNT_ONE)
+        userDepositsBeforeSecondWithdrawal
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/CollateralDepositRecord.test.ts:150]: move to separate test

File: apps/smart-contracts/core/test/CollateralDepositRecord.test.ts [hunk 4]
```diff
@@ -119,51 +120,47 @@ describe('=> CollateralDepositRecord', () => {
...
 
       await depositRecord.connect(user).recordWithdrawal(user.address, TEST_AMOUNT_TWO)
 
       expect(await depositRecord.getGlobalNetDepositAmount()).to.eq(
         globalDepositsBefore.sub(TEST_AMOUNT_TWO)
       )
-      expect(await depositRecord.getUserDepositAmount(user.address)).to.eq(
-        accountDepositsBefore.sub(TEST_AMOUNT_TWO)
-      )
+      expect(await depositRecord.getUserDepositAmount(user.address)).to.eq(userDepositsBefore)
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/CollateralDepositRecord.test.ts:134]: move to separate test

File: apps/smart-contracts/core/test/CollateralDepositRecord.test.ts [hunk 5]
```diff
@@ -119,51 +120,47 @@ describe('=> CollateralDepositRecord', () => {
       ).to.be.revertedWith('msg.sender != allowed hook')
     })
 
-    it("should correctly subtract 'amount' from both deposited totals when starting from a non-zero value", async () => {
-      expect(await depositRecord.getGlobalNetDepositAmount()).to.be.gt(0)
-      expect(await depositRecord.getUserDepositAmount(user.address)).to.be.gt(0)
+    it("should correctly subtract 'amount' from global total when starting from a non-zero value", async () => {
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/CollateralDepositRecord.test.ts:123]: there seems to be missing tests for when subtracting 0 from a non-zero value and subtracting 0 from a zero value (that no balances change)

File: apps/smart-contracts/core/test/CollateralDepositRecord.test.ts [hunk 6]
```diff
@@ -119,51 +120,47 @@ describe('=> CollateralDepositRecord', () => {
...
       )
-      expect(await depositRecord.getUserDepositAmount(user.address)).to.eq(
-        accountDepositsBefore.sub(TEST_AMOUNT_TWO)
-      )
+      expect(await depositRecord.getUserDepositAmount(user.address)).to.eq(userDepositsBefore)
     })
 
-    it("should correctly subtract 'amount' from both deposited totals when called again", async () => {
-      await depositRecord.connect(user).recordWithdrawal(user.address, TEST_AMOUNT_TWO)
+    it("should correctly subtract 'amount' from global total when called again", async () => {
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/CollateralDepositRecord.test.ts:137]:
```suggestion
    it("subtracts from global deposits if called again", async () => {
```

File: apps/smart-contracts/core/test/CollateralDepositRecord.test.ts [hunk 7]
```diff
@@ -106,9 +106,10 @@ describe('=> CollateralDepositRecord', () => {
...
   describe('# recordWithdrawal', () => {
-    const testDepositToWithdrawFrom = parseEther('5')
     beforeEach(async () => {
-      await depositRecord.connect(user).recordDeposit(user.address, testDepositToWithdrawFrom)
+      await depositRecord
+        .connect(user)
+        .recordDeposit(user.address, TEST_AMOUNT_ONE.add(TEST_AMOUNT_TWO))
     })
 
     it('should only be callable by allowed contracts', async () => {
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/CollateralDepositRecord.test.ts:115]: missing a test for withdrawing from an address that never deposited (e.g. zero address)

Comment (XavierEkkel) [reply apps/smart-contracts/core/test/CollateralDepositRecord.test.ts:115]: actually, isn't the address argument now unnecessary for withdrawals given that user deposits aren't net?

Comment (ramenforbreakfast) [reply apps/smart-contracts/core/test/CollateralDepositRecord.test.ts:115]: You are right, I am removing it now.

**Comment by ramenforbreakfast** [comment]
I have addressed the comments with the following changes:
- Remove "should" wording and usage of `amount` inside testcase names to conform to our current testcase nomenclature
- Added a testcases for withdrawing 0 when a balance is already 0
- Separated out global and user deposit checks into different tests
- Change usage of `global total` to `global net deposits`
- Simplified names of testcases testing subsequent withdrawals to `if called again`
- Replaced `when` wording with `if` to be consistent with other test suites.

I have only applied these changes to the tests that pertain to this change. We can retroactively rename other parts of the test suite if time allows later.

**Comment by ramenforbreakfast** [comment]
Rebased to reflect changes introduced from PR #459-462

File: apps/smart-contracts/core/contracts/WithdrawHook.sol [hunk 8]
```diff
@@ -58,7 +58,7 @@ contract WithdrawHook is IWithdrawHook, SafeAccessControlEnumerable {
       );
       _globalAmountWithdrawnThisPeriod += _amountBeforeFee;
     }
-    _depositRecord.recordWithdrawal(_sender, _amountBeforeFee);
+    _depositRecord.recordWithdrawal(_amountBeforeFee);
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/contracts/WithdrawHook.sol:61]: lets add a natspec dev comment (in this implementation contract only) that we use amountBeforeFee to better reflect TVL changes and that amountAfterFee is ignored

File: apps/smart-contracts/core/test/CollateralDepositRecord.test.ts [hunk 9]
```diff
@@ -106,64 +106,122 @@ describe('=> CollateralDepositRecord', () => {
...
       )
-      expect(await depositRecord.getUserDepositAmount(user.address)).to.eq(
-        accountDepositsBefore.sub(TEST_AMOUNT_TWO)
-      )
     })
 
-    it("should correctly subtract 'amount' from both deposited totals when called again", async () => {
-      await depositRecord.connect(user).recordWithdrawal(user.address, TEST_AMOUNT_TWO)
+    it('leaves user deposits unchanged if withdrawal > 0 and user deposits > 0', async () => {
+      const userDepositsBefore = await depositRecord.getUserDepositAmount(user.address)
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/CollateralDepositRecord.test.ts:135]: nit: same for other tests
```suggestion
    it('leaves user deposits unchanged if withdrawal > 0 and user deposit > 0', async () => {
      const userDepositBefore = await depositRecord.getUserDepositAmount(user.address)
```

**Review by XavierEkkel — APPROVED** [review]
LGTM after nits
